### PR TITLE
[MDB Ignore] Refactoring Flora code

### PR DESCRIPTION
**Pull request by DragonTrance**
## About The Pull Request

Refactors most of flora code. Instead of only trees, rocks, and lavaland flora using their own code, I've made it so all the flora shares the same concept of having the ability to harvest plants. Trees and rocks give a random number of logs/sand when harvested now, but that can be easily changed. Trees also require an axe, hatchet, or e-sword to cut down, instead of assuming anything sharp can cut it.

I've also added the ability to "uproot" flora with shovels, so flora can be moved around.

Some typepaths have been renamed, and 99% of flora in maps use a different typepath now

## Why It's Good For The Game
Flora code was tossed around everywhere that made you think "what the fuck". I've organized it

## Changelog

:cl:
refactor: Flora code is much more organized :)
add: Flora (trees, plants, etc) can be uprooted with a shovel, which can then be dragged around.
code: Moved "kirbyplants" into their own .dm file, instead of just finding a comfortable spot in flora code
/:cl:

### CODE REVIEW

**Comment by san7890** [comment]
Put `[MDB Ignore]` in your title (stands for MapDiffBot Ignore) so we don't burn down AffectedArc's servers for the Original MapDiffBot.

File: code/modules/awaymissions/mission_code/snowdin.dm [hunk 1]
```diff
@@ -473,13 +473,7 @@
...
 //objs//--
 
-/obj/structure/flora/rock/icy
-	name = "icy rock"
-	color = rgb(204,233,235)
-
-/obj/structure/flora/rock/pile/icy
-	name = "icey rocks"
-	color = rgb(204,233,235)
+//ice rocks moved to flora.dm
```
Comment (Fikou) [inline code/modules/awaymissions/mission_code/snowdin.dm:476]: dont make comments like that, they dont describe the current form of the code

File: code/game/objects/structures/flora.dm [hunk 2]
```diff
@@ -1,83 +1,411 @@
...
+	
+	/// If false, the flora won't be able to be harvested at all. If it's true, go through checks normally to determine if the flora is able to be harvested
+	var/harvestable = TRUE
+	/// The low and high ends of how many product_type items you get
+	var/harvest_amount_low = 1
+	var/harvest_amount_high = 3
+	/// Messages to show to the user when the flora gets harvested
+	var/harvest_message_low as text
+	var/harvest_message_med as text
+	var/harvest_message_high as text
```
Comment (Fikou) [inline code/game/objects/structures/flora.dm:45]: we dont really use as for variables

File: code/game/objects/structures/flora.dm [hunk 3]
```diff
@@ -1,83 +1,411 @@
...
+	var/uprooting_tools = list(/obj/item/shovel)
+	var/uprooted = FALSE
+	var/previous_rotation = 0
+	
+	/// If false, the flora won't be able to be harvested at all. If it's true, go through checks normally to determine if the flora is able to be harvested
+	var/harvestable = TRUE
+	/// The low and high ends of how many product_type items you get
+	var/harvest_amount_low = 1
+	var/harvest_amount_high = 3
+	/// Messages to show to the user when the flora gets harvested
```
Comment (Fikou) [inline code/game/objects/structures/flora.dm:42]: you arent using dmdoc correctly, dont use one comment for multiple lines

File: code/game/objects/items/kirbyplants.dm [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	throwforce = 13
+	throw_speed = 2
+	throw_range = 4
+	item_flags = NO_PIXEL_RANDOM_DROP
+
+	/// Can this plant be trimmed by someone with TRAIT_BONSAI
+	var/trimmable = TRUE
+	var/list/static/random_plant_states
+
+/obj/item/kirbyplants/ComponentInitialize()
```
Comment (Kylerace) [inline code/game/objects/items/kirbyplants.dm:20]: ComponentInitialize() is cringe dont use it for new things. use Initialize() instead

File: code/game/objects/structures/flora.dm [hunk 5]
```diff
@@ -1,83 +1,413 @@
...
+	if(!required_tools)
+		required_tools = list()
+		if(flora_flags & FLORA_WOODEN)
+			required_tools += FLORA_HARVEST_WOOD_TOOLS
+		if(flora_flags & FLORA_STONE)
+			required_tools += FLORA_HARVEST_STONE_TOOLS
+	
+	required_tools = typecacheof(required_tools)
+	disallowed_tools = typecacheof(disallowed_tools)
+	uprooting_tools = typecacheof(uprooting_tools)
```
Comment (Kylerace) [inline code/game/objects/structures/flora.dm:75]: use a global list for these so you only have to do this once for each unique subtype please, id rather not deal with this type of performance problem again

File: code/game/objects/structures/flora.dm [hunk 6]
```diff
@@ -1,83 +1,413 @@
...
+		use_default_sound = FALSE
+	if(use_default_sound)
+		return ..()
+
+/* 
+ * A helper proc for getting a random amount of products, associated with the flora's product list.
+ * Returns: A list where each value is (product_type = amount_of_products)
+ */
+/obj/structure/flora/proc/get_products_list()
+	. = list()
```
Comment (Kylerace) [inline code/game/objects/structures/flora.dm:155]: use a named list that you return please

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/game/objects/items/kirbyplants.dm (hunk 4 above)
Comment (DragonTrance) [reply code/game/objects/items/kirbyplants.dm:20]: I moved the old code into a new file, I touched nothing of that. These are the potted plants you see throughout the station

**Pull request merged**